### PR TITLE
cql3: replace cql3::selection::selectable::raw hierarchy with expressions

### PR DESCRIPTION
**Pull request by avikivity**
Currently, the grammar has two parallel hierarchies. One hierarchy is
used in the WHERE clause, and is based on a combination of `term`
and expressions. The other is used in the SELECT clause, and is
using the cql3::selection::selectable hierarchy. There is some overlap
between the hierarchies: both can name columns. Logically, however,
they overlap completely - in SQL anything you can select you can
filter on, and vice versa. So merging the two hierarchies is important if
we want to enrich CQL. This series does that, partially (see below),
converting the SELECT clause to expressions.

There is another hierarchy split: between the "raw", pre-prepare object
hierarchy, and post-prepare non-raw. This series limits itself to converting
the raw hierarchy and leaves the non-raw hierarchy alone.

An important design choice is not to have this raw/non-raw split in expressions.
Note that most of the hierarchy is completely parallel: addition is addition
both before prepare and after prepare (but see [1]). The main difference
is around identifiers - before preparation they are unresolved, and after
preparation they become `column_definition` objects. We resolve that by
having two separate types: `unresolved_identifier` for the pre-prepare phase,
and the existing `column_value` for post-prepare phase.

Alternative choices would be to keep a separate expression::raw variant, or
to template the expression variant on whether it is raw or not. I think it would
cause undue bloat and confusion.

Note the series introduces many on_internal_error() calls. This is because
there is not a lot of overlap in the hierarchies today; you can't have a cast in
the WHERE clause, for example. These on_internal_error() calls cannot be
triggered since the grammar does not yet allow such expressions to be
expressed. As we expand the grammar, they will have to be replaced with
working implementations.

Lastly, field selection is expressible in both hierarchies. This series does not yet
merge the two representations (`column_value.sub` vs `field_selection`), but it
should be easy to do so later.

[1] the `+` operator can also be translated to list concatenation, which we may
  choose to represent by yet another type.

Test: unit(dev)

### CODE REVIEW

**Comment by avikivity** [comment]
v2: I forgot to properly changelog the first patch.

File: cql3/expr/expression.hh [hunk 1]
```diff
@@ -72,6 +72,23 @@ class unresolved_identifier;
 using expression = std::variant<bool, conjunction, binary_operator, column_value, column_value_tuple, token,
                                 unresolved_identifier>;
 
+// An expression variant element can't contain an expression by value, since the size of the variant
+// will be infinite. `nested_expression` contains an expression indirectly, but has value semantics and
+// is copyable.
+class nested_expression {
```
Comment (nyh) [inline cql3/expr/expression.hh:106]: nitpick: I think maybe the names `indirect_expression` or `expression_ptr` would be more self-explanatory than `nested_expression`. But I guess `nested_expression` isn't that bad, so I'll accept.

File: cql3/expr/expression.cc [hunk 2]
```diff
@@ -53,6 +53,22 @@ using boost::adaptors::filtered;
...
+        : _e(std::make_unique<expression>(std::move(e))) {
+}
+
+nested_expression::nested_expression(const nested_expression& o)
+        : nested_expression(*o._e) {
+}
+
+nested_expression&
+nested_expression::operator=(const nested_expression& o) {
+    if (this != &o) {
```
Comment (nyh) [inline cql3/expr/expression.cc:66]: nitpick: it seems to me that this if() is not necessary - the code will be correct (just waste a little time) even without it. But it's not wrong, so ok.

File: cql3/statements/create_view_statement.cc [hunk 3]
```diff
@@ -212,6 +212,12 @@ future<shared_ptr<cql_transport::event::schema_change>> create_view_statement::a
         if (dynamic_pointer_cast<selection::selectable::writetime_or_ttl::raw>(selectable)) {
             throw exceptions::invalid_request_exception(format("Cannot use function when defining a materialized view"));
         }
+        if (auto e = dynamic_pointer_cast<selection::selectable::with_expression::raw>(selectable)) {
```
Comment (psarna) [inline cql3/statements/create_view_statement.cc]: Looks like this change got into a wrong patch? It doesn't seem related to making column_identifier::raw forward declarable.

File: cql3/expr/expression.cc [hunk 4]
```diff
@@ -900,6 +1006,11 @@ expression replace_column_def(const expression& expr, const column_definition* n
                 throw std::logic_error(format("replace_column_def invalid with column tuple: {}", to_string(expr)));
             },
             [&] (const token&) { return expr; },
+            [&] (const unresolved_identifier&) { return expr; },
+            [&] (const column_mutation_attribute&) { return expr; },
+            [&] (const function_call&) { return expr; },
+            [&] (const cast&) { return expr; },
+            [&] (const field_selection&) { return expr; },
```
Comment (dekimir) [inline cql3/expr/expression.cc:1013]: In the fullness of time, we'll want to descend into nested expressions here, right?  If so, please add a TODO comment.  If not, then the function's blurb should mention that limitation.

File: cql3/expr/expression.cc [hunk 5]
```diff
@@ -864,6 +946,30 @@ std::ostream& operator<<(std::ostream& os, const expression& expr) {
...
+                    [&] (const shared_ptr<functions::function>& anon) {
+                        fmt::print(os, "<anonymous function>({}})", fmt::join(fc.args, ", "));
+                    },
+                }, fc.func);
+            },
+            [&] (const cast& c)  {
+                fmt::print(os, "({}} AS {})", *c.arg, c.type);
+            },
+            [&] (const field_selection& fs)  {
+                fmt::print(os, "({}}.{})", *fs.structure, fs.field);
```
Comment (dekimir) [inline cql3/expr/expression.cc:971]: There are more `}`s than `{`s in the format string -- is that intentional?  What does it mean?

File: cql3/restrictions/statement_restrictions.cc [hunk 6]
```diff
@@ -210,6 +210,10 @@ static std::vector<expr::expression> extract_partition_range(
         }
 
         void operator()(bool) {}
+
+        void operator()(const unresolved_identifier&) {
+            on_internal_error(rlogger, "extract_partition_range(unresolved_identifier");
```
Comment (psarna) [inline cql3/restrictions/statement_restrictions.cc:215]: Is the unmatched opening bracket intentional? It looks weird

File: cql3/Cql.g [hunk 7]
```diff
@@ -441,30 +441,25 @@ selectClause returns [std::vector<shared_ptr<raw_selector>> expr]
...
-                                                                                              cql3::expr::unresolved_identifier{std::move(c)}}); }
-       | K_TTL       '(' c=cident ')'              { tmp = make_shared<selectable::with_expression::raw>(
-                                                        cql3::expr::column_mutation_attribute{cql3::expr::column_mutation_attribute::attribute_kind::ttl,
-                                                                                              cql3::expr::unresolved_identifier{std::move(c)}}); }
-       | f=functionName args=selectionFunctionArgs { tmp = ::make_shared<selectable::with_expression::raw>(
-                                                        cql3::expr::function_call{std::move(f), std::move(args)}); }
-       | K_CAST      '(' arg=unaliasedSelector K_AS t=native_type ')'  { tmp = ::make_shared<selectable::with_expression::raw>(
-                                                        cql3::expr::cast{std::move(arg), std::move(t)}); }
+    : us=unaliasedSelector (K_AS c=ident { alias = c; })? { $s = ::make_shared<raw_selector>(us, alias); }
```
Comment (nyh) [inline cql3/Cql.g:444]: Unnecessary (?), but not wrong, change?

File: cql3/restrictions/statement_restrictions.cc [hunk 8]
```diff
@@ -275,6 +279,10 @@ static std::vector<expr::expression> extract_clustering_prefix_restrictions(
         }
 
         void operator()(bool) {}
+
+        void operator()(const unresolved_identifier&) {
+            on_internal_error(rlogger, "extract_clustering_prefix_restrictions(unresolved_identifier");
```
Comment (psarna) [inline cql3/restrictions/statement_restrictions.cc:300]: ditto - bracket

**Review by nyh — APPROVED** [review]

**Comment by nyh** [comment]
@dekimir I think this series as good and am ready to commit it, but maybe you want to review it too?

File: cql3/expr/expression.hh [hunk 9]
```diff
@@ -72,13 +72,29 @@ class unresolved_identifier;
...
+concept expression_element = is_expression_element<T>::value;
+
 // An expression variant element can't contain an expression by value, since the size of the variant
 // will be infinite. `nested_expression` contains an expression indirectly, but has value semantics and
 // is copyable.
 class nested_expression {
     std::unique_ptr<expression> _e;
 public:
     nested_expression(expression e);
+    nested_expression(expression_element auto e);
```
Comment (psarna) [inline cql3/expr/expression.hh:110]: Just to make sure I understand: is `expression_element auto e` syntactic sugar for "this is actually a template parameter restricted by the `expression_element` concept"?

File: cql3/Cql.g [hunk 10]
```diff
@@ -447,14 +447,15 @@ selector returns [shared_ptr<raw_selector> s]
 unaliasedSelector returns [shared_ptr<selectable::raw> s]
     @init { shared_ptr<selectable::raw> tmp; }
     :  ( c=cident                                  { tmp = make_shared<selectable::with_expression::raw>(cql3::expr::unresolved_identifier{std::move(c)}); }
-       | K_COUNT '(' countArgument ')'             { tmp = selectable::with_function::raw::make_count_rows_function(); }
+       | K_COUNT '(' countArgument ')'             { tmp = ::make_shared<selectable::with_expression::raw>(make_count_rows_function_expression()); }
```
Comment (psarna) [inline cql3/Cql.g]: (not a review) /cc @havaker - keep track of these changes, it's probably going to change the way we can recognize if a request is an aggregate

**Review by psarna — COMMENTED** [review]
Nice.

**Review by dekimir — COMMENTED** [review]
> @dekimir I think this series as good and am ready to commit it, but maybe you want to review it too?

Yes, please hold it until I finish my review.

File: cql3/statements/create_view_statement.cc (hunk 3 above)
Comment (avikivity) [reply cql3/statements/create_view_statement.cc]: Yes, it belongs to the patch introducing with_expression.

File: cql3/expr/expression.hh [hunk 11]
```diff
@@ -65,9 +74,49 @@ class conjunction;
 struct column_value;
 struct column_value_tuple;
 struct token;
+class unresolved_identifier;
+class column_mutation_attribute;
+class function_call;
+class cast;
+class field_selection;
 
 /// A restriction expression -- union of all possible restriction types.  bool means a Boolean constant.
```
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:83]: yes

File: cql3/expr/expression.hh [hunk 12]
```diff
@@ -106,24 +155,50 @@ enum class comparison_order : char {
...
 };
 
 /// A conjunction of restrictions.
 struct conjunction {
     std::vector<expression> children;
 };
 
+// Gets resolved eventually into a column_value.
+struct unresolved_identifier {
+    shared_ptr<column_identifier_raw> ident;
```
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:173]: I'm sure it's not everywhere.

File: cql3/restrictions/statement_restrictions.cc (hunk 6 above)
Comment (avikivity) [reply cql3/restrictions/statement_restrictions.cc:215]: No, accidental, will fix before everyone's brain melts

File: cql3/expr/expression.hh [hunk 13]
```diff
@@ -65,9 +74,49 @@ class conjunction;
...
+};
+
+template <typename T, typename U, typename... Elements>
+struct is_expression_element<T, std::variant<U, Elements...>> : is_expression_element<T, std::variant<Elements...>>{
+};
+
+template <typename T>
+concept expression_element = is_expression_element<T>::value;
+
+// An expression variant element can't contain an expression by value, since the size of the variant
```
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:103]: It's just a dereference, it's not really a new type.

Another way to go is to make expression a struct and make the variant a unique_ptr in that struct. I think we'll need to go there in order to annotate expressions with types, but so far the big variant is fine.

File: cql3/expr/expression.hh (hunk 1 above)
Comment (avikivity) [reply cql3/expr/expression.hh:106]: expression_ptr implies pointer semantics (copying the pointer does not copy the pointee)

File: cql3/expr/expression.hh [hunk 14]
```diff
@@ -65,9 +74,49 @@ class conjunction;
...
+class cast;
+class field_selection;
 
 /// A restriction expression -- union of all possible restriction types.  bool means a Boolean constant.
-using expression = std::variant<bool, conjunction, binary_operator, column_value, column_value_tuple, token>;
+using expression = std::variant<bool, conjunction, binary_operator, column_value, column_value_tuple, token,
+                                unresolved_identifier, column_mutation_attribute, function_call, cast,
+                                field_selection>;
+
+template <typename T, typename Variant = expression>
```
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:88]: The template accepts a variant, and the default is the variant that is `expression`. However, when processing, it's instantiated for other variants.

I'll probably divorce with from expressions and make it a utility, not because I expect reuse, but because readers won't suffer from mixing of abstraction levels.

File: cql3/expr/expression.hh [hunk 15]
```diff
@@ -65,9 +74,49 @@ class conjunction;
...
+template <typename T, typename... Elements>
+struct is_expression_element<T, std::variant<T, Elements...>> : std::true_type {
+};
+
+template <typename T, typename U, typename... Elements>
+struct is_expression_element<T, std::variant<U, Elements...>> : is_expression_element<T, std::variant<Elements...>>{
+};
+
+template <typename T>
+concept expression_element = is_expression_element<T>::value;
```
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:101]: Nice. I'll drop the whole thing in favor of this.

File: cql3/expr/expression.hh [hunk 16]
```diff
@@ -65,9 +74,49 @@ class conjunction;
...
+
+// An expression variant element can't contain an expression by value, since the size of the variant
+// will be infinite. `nested_expression` contains an expression indirectly, but has value semantics and
+// is copyable.
+class nested_expression {
+    std::unique_ptr<expression> _e;
+public:
+    nested_expression(expression e);
+    nested_expression(expression_element auto e);
```
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:110]: Yes, explicit should be the default and deviating from it should require a comment.

File: cql3/expr/expression.hh (hunk 9 above)
Comment (avikivity) [reply cql3/expr/expression.hh:110]: Yes, see https://en.cppreference.com/w/cpp/language/function_template#Abbreviated_function_template

File: cql3/expr/expression.hh (hunk 14 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:88]: Update: I'll divorce it from existing.

File: cql3/expr/expression.hh [hunk 17]
```diff
@@ -106,24 +155,50 @@ enum class comparison_order : char {
...
+
+    ~unresolved_identifier();
+};
+
+// An attribute attached to a column mutation: writetime or ttl
+struct column_mutation_attribute {
+    enum class attribute_kind { writetime, ttl };
+
+    attribute_kind kind;
+    // note: unresolved_identifier is legal here now, later column_value.
```
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:183]: A bit of both. It's legal today, before prepare. Tomorrow, we'll implement preparing unresolved_identifier into an column_value and it will be legal after prepare().

File: cql3/Cql.g (hunk 7 above)
Comment (avikivity) [reply cql3/Cql.g:444]: It is necessary. `us` became an std::variant, so argument-dependent lookup now finds std::make_shared since one of the arguments is in the std namespace.

File: cql3/expr/expression.cc (hunk 5 above)
Comment (avikivity) [reply cql3/expr/expression.cc:971]: Just a bug.

**Comment by avikivity** [comment]
> @dekimir I think this series as good and am ready to commit it, but maybe you want to review it too?

I'll need a second pass.

File: cql3/expr/expression.hh (hunk 12 above)
[the remark this answers is not in the document]
Comment (dekimir) [reply cql3/expr/expression.hh:173]: I meant everywhere in this file.

File: cql3/expr/expression.hh (hunk 17 above)
[the remark this answers is not in the document]
Comment (dekimir) [reply cql3/expr/expression.hh:183]: I like this explanation and wish I read it in the comment. :)

File: cql3/expr/expression.cc (hunk 5 above)
Comment (dekimir) [reply cql3/expr/expression.cc:971]: Note it also appears in other cases above.

File: cql3/expr/expression.hh (hunk 15 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:101]: Unfortunately, doesn't work. std::get<T>(e) requires all the member types of expression to be defined (not just forward declared). expression_element is required to be defined before nested_expression. And nested_expression needs to be defined before the expression elements that use it. So we have a loop.

[the remark this answers is not in the document]
Comment (dekimir) [reply cql3/expr/expression.hh:101]: > `std::get<T>(e) `requires all the member types of expression to be defined (not just forward declared)

Does it? https://godbolt.org/z/TqKvTf7T4

[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:101]: Somehow f1() compiles for non-members too. https://godbolt.org/z/a1xK161xb

[the remark this answers is not in the document]
Comment (dekimir) [reply cql3/expr/expression.hh:101]: Weird; must be a bug (in both gcc and clang).

If I invoke `std::get<Q>` explicitly on Var, compilation fails: https://godbolt.org/z/KbqbGsoEx

[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:101]: Looks like std::get<> uses static_assert instead of a constraint, so in the context of a constrait std::get<non_member>(v) passes.

File: cql3/expr/expression.cc (hunk 4 above)
Comment (avikivity) [reply cql3/expr/expression.cc:1013]: I'll mention it in the patch changelogs for completeness.

Comment (avikivity) [reply cql3/expr/expression.cc:1013]: Ah, with_function_call already did, I'll update the others.

File: cql3/expr/expression.cc (hunk 5 above)
Comment (avikivity) [reply cql3/expr/expression.cc:971]: At least I'm consistent

File: cql3/expr/expression.hh [hunk 18]
```diff
@@ -216,6 +307,17 @@ size_t count_if(const expression& e, Fn f) {
             [] (const column_value&) -> size_t { return 0; },
             [] (const column_value_tuple&) -> size_t { return 0; },
             [] (const token&) -> size_t { return 0; },
+            [] (const unresolved_identifier&) -> size_t { return 0; },
+            [] (const column_mutation_attribute&) -> size_t { return 0; },
+            [&] (const function_call& fc) -> size_t {
+                return std::accumulate(fc.args.cbegin(), fc.args.cend(), size_t{0},
+                                       [&] (size_t acc, const expression& c) { return acc + count_if(c, f); });
+            },
```
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:315]: Ho hum. The form is exactly the same, but the meaning is completely different. One a members of an AND, the other is a list of function arguments.

File: cql3/selection/selectable.cc [hunk 19]
```diff
@@ -191,14 +138,114 @@ selectable::with_cast::to_string() const {
 }
 
 shared_ptr<selectable>
-selectable::with_cast::raw::prepare(const schema& s) const {
-    return ::make_shared<selectable::with_cast>(_arg->prepare(s), _type);
+prepare_selectable(const schema& s, const expr::expression& raw_selectable) {
+    return std::visit(overloaded_functor{
+        [&] (bool bool_constant) -> shared_ptr<selectable> {
+            throw std::runtime_error("no way to express SELECT TRUE/FALSE in the grammar yet");
```
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/selection/selectable.cc:144]: I'll change it. I've been thinking as I write it that internal error is exaggerating things, but a patch series converting stuff is not the place to think it.

[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/selection/selectable.cc:144]: Fixed now.

File: cql3/statements/create_view_statement.cc (hunk 3 above)
Comment (avikivity) [reply cql3/statements/create_view_statement.cc]: Fixed. Also moved the definition of with_expression::raw::expression().

File: cql3/expr/expression.hh (hunk 12 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:173]: Fixed.

File: cql3/restrictions/statement_restrictions.cc (hunk 6 above)
Comment (avikivity) [reply cql3/restrictions/statement_restrictions.cc:215]: Fixed.

File: cql3/restrictions/statement_restrictions.cc (hunk 8 above)
Comment (avikivity) [reply cql3/restrictions/statement_restrictions.cc:300]: ditto

File: cql3/expr/expression.hh [hunk 20]
```diff
@@ -65,9 +74,49 @@ class conjunction;
...
+class field_selection;
 
 /// A restriction expression -- union of all possible restriction types.  bool means a Boolean constant.
-using expression = std::variant<bool, conjunction, binary_operator, column_value, column_value_tuple, token>;
+using expression = std::variant<bool, conjunction, binary_operator, column_value, column_value_tuple, token,
+                                unresolved_identifier, column_mutation_attribute, function_call, cast,
+                                field_selection>;
+
+template <typename T, typename Variant = expression>
+struct is_expression_element : std::false_type {
```
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:89]: It's now in utils/variant_element.hh.

File: cql3/expr/expression.hh (hunk 16 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:110]: Added a comment.

File: cql3/expr/expression.hh (hunk 11 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:83]: done

File: cql3/expr/expression.hh (hunk 17 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:183]: added

**Comment by avikivity** [comment]
v3:
- move stub handling of with_expression::raw in create_view_statement.cc to patch that introduces with_expression
- introduce new utility concept variant_element<> instead of defining it in expression.hh
- explain non-explicit constructors
- fix types in string literals in on_internal_error() and format() calls
- explain patches that don't fully print the expressions they introduce
- replace plain throw with on_internal_error when we think it can't happen
- improved comment explaining why column_mutation_attribute only takes an unresolved_identifier
- change shared_ptr to ::shared_ptr in expression.hh unresolved_identifier definition
- updated expression's blurb to express that it's no longer restricted to restrictions

**Review by nyh — APPROVED** [review]
Looks good, but I left just one style nitpick - that our usual coding convention is that concept names should be capitalized.

File: cql3/expr/expression.hh [hunk 21]
```diff
@@ -72,6 +73,9 @@ class unresolved_identifier;
 using expression = std::variant<bool, conjunction, binary_operator, column_value, column_value_tuple, token,
                                 unresolved_identifier>;
 
+template <typename T>
+concept expression_element = utils::variant_element<T, expression>;
```
[the remark this answers is not in the document]
Comment (nyh) [reply cql3/expr/expression.hh:90]: Correction: if you are to believe what you yourself wrote in the Seastar coding-style.md,
> Template parameters and concepts use `CamelCase`

So it needs to be VariantElement, ExpressionElement.

**Comment by avikivity** [comment]
> Looks good, but I left just one style nitpick - that our usual coding convention is that concept names should be capitalized.

That was brought in from the concepts technical specification. The standard later reverted to snake_case. We have plenty of snake_case.

File: cql3/expr/expression.hh (hunk 21 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh:90]: I'll switch it to CamelCase to avoid a style flamewar, and attempt to adjust the coding style independently.

**Comment by avikivity** [comment]
v4:
 - switch concept names to CamelCase

File: cql3/expr/expression.cc (hunk 4 above)
Comment (dekimir) [reply cql3/expr/expression.cc:1013]: > The next-to-last patch does that.

Does what?  Does it add TODOs here or modify the blurb in expression.hh?

**Comment by nyh** [comment]
> > Looks good, but I left just one style nitpick - that our usual coding convention is that concept names should be capitalized.
> 
> That was brought in from the concepts technical specification. The standard later reverted to snake_case. We have plenty of snake_case.

Holy shit, I didn't notice that this happened:
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1754r1.pdf
http://open-std.org/JTC1/SC22/WG21/docs/papers/2019/p1851r0.html

We need to update the Seastar coding style and concept names.

**Comment by nyh** [comment]
Unfortunately, this doesn't build on my Fedora 34 with clang 12.0.0., I get errors:

```
[31/205] CXX build/dev/cql3/expr/expression.o
FAILED: build/dev/cql3/expr/expression.o 
clang++ -MD -MT build/dev/cql3/expr/expression.o -MF build/dev/cql3/expr/expression.o.d -I/home/nyh/scylla.mai
ntainer/seastar/include -I/home/nyh/scylla.maintainer/build/dev/seastar/gen/include -std=gnu++20 -U_FORTIFY_SO
URCE -DSEASTAR_SSTRING -Werror=unused-result -fstack-clash-protection -DSEASTAR_API_LEVEL=6 -DSEASTAR_ENABLE_A
LLOC_FAILURE_INJECTION -DSEASTAR_TYPE_ERASE_MORE -DFMT_LOCALE -DFMT_SHARED -I/usr/include/p11-kit-1  -DDEVEL -
DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSCYLLA_ENABLE_ERROR_INJECTION -O2 -iquote. -iquote build/dev/gen --s
td=gnu++20  -ffile-prefix-map=/home/nyh/scylla.maintainer=.  -march=westmere -DBOOST_TEST_DYN_LINK   -Iabseil 
-fvisibility=hidden  -Wall -Werror -Wno-mismatched-tags -Wno-tautological-compare -Wno-parentheses-equality -W
no-c++11-narrowing -Wno-sometimes-uninitialized -Wno-return-stack-address -Wno-missing-braces -Wno-unused-lamb
da-capture -Wno-overflow -Wno-noexcept-type -Wno-error=cpp -Wno-ignored-attributes -Wno-overloaded-virtual -Wn
o-unused-command-line-argument -Wno-inconsistent-missing-override -Wno-defaulted-function-deleted -Wno-redecla
red-class-member -Wno-unsupported-friend -Wno-unused-variable -Wno-delete-non-abstract-non-virtual-dtor -Wno-u
nknown-attributes -Wno-braced-scalar-init -Wno-implicit-int-float-conversion -Wno-delete-abstract-non-virtual-
dtor -Wno-uninitialized-const-reference -Wno-psabi -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEAS
TAR_TESTING_MAIN -DHAVE_LZ4_COMPRESS_DEFAULT  -c -o build/dev/cql3/expr/expression.o cql3/expr/expression.cc
In file included from cql3/expr/expression.cc:22:
In file included from ./cql3/expr/expression.hh:24:
In file included from /usr/include/fmt/core.h:13:
In file included from /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/functional:49:
In file included from /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/stl_function.h:60:
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/type_traits:2933:5: error: no type named 'type' in 'std::invoke_result<overloaded_functor<(lambda at cql3/expr/expression.cc:1019:13), (lambda at cql3/expr/expression.cc:1020:13), (lambda at cql3/expr/expression.cc:1025:13), (lambda at cql3/expr/expression.cc:1028:13), (lambda at cql3/expr/expression.cc:1031:13), (lambda at cql3/expr/expression.cc:1034:13)>, const cql3::expr::unresolved_identifier &>'
    using invoke_result_t = typename invoke_result<_Fn, _Args...>::type;
    ^~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/variant:1081:2: note: in instantiation of template type alias 'invoke_result_t' requested here
        invoke_result_t<_Visitor, __get_t<_Idxs, _Variant>>...
        ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/variant:1753:6: note: in instantiation of function template specialization 'std::__detail::__variant::__check_visitor_results<overloaded_functor<(lambda at cql3/expr/expression.cc:1019:13), (lambda at cql3/expr/expression.cc:1020:13), (lambda at cql3/expr/expression.cc:1025:13), (lambda at cql3/expr/expression.cc:1028:13), (lambda at cql3/expr/expression.cc:1031:13), (lambda at cql3/expr/expression.cc:1034:13)>, const std::variant<bool, cql3::expr::conjunction, cql3::expr::binary_operator, cql3::expr::column_value, cql3::expr::column_value_tuple, cql3::expr::token, cql3::expr::unresolved_identifier, cql3::expr::column_mutation_attribute, cql3::expr::function_call, cql3::expr::cast, cql3::expr::field_selection> &, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10>' requested here
            __check_visitor_results<_Visitor, _Variants...>(
            ^
cql3/expr/expression.cc:1018:17: note: in instantiation of function template specialization 'std::visit<overloaded_functor<(lambda at cql3/expr/expression.cc:1019:13), (lambda at cql3/expr/expression.cc:1020:13), (lambda at cql3/expr/expression.cc:1025:13), (lambda at cql3/expr/expression.cc:1028:13), (lambda at cql3/expr/expression.cc:1031:13), (lambda at cql3/expr/expression.cc:1034:13)>, const std::variant<bool, cql3::expr::conjunction, cql3::expr::binary_operator, cql3::expr::column_value, cql3::expr::column_value_tuple, cql3::expr::token, cql3::expr::unresolved_identifier, cql3::expr::column_mutation_attribute, cql3::expr::function_call, cql3::expr::cast, cql3::expr::field_selection> &>' requested here
    return std::visit(overloaded_functor{
                ^
In file included from cql3/expr/expression.cc:22:
In file included from ./cql3/expr/expression.hh:27:
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/variant:1752:44: error: constexpr variable '__visit_rettypes_match' must be initialized by a constant expression
          constexpr bool __visit_rettypes_match = __detail::__variant::
                                                  ^~~~~~~~~~~~~~~~~~~~~
cql3/expr/expression.cc:1018:17: note: in instantiation of function template specialization 'std::visit<overloaded_functor<(lambda at cql3/expr/expression.cc:1019:13), (lambda at cql3/expr/expression.cc:1020:13), (lambda at cql3/expr/expression.cc:1025:13), (lambda at cql3/expr/expression.cc:1028:13), (lambda at cql3/expr/expression.cc:1031:13), (lambda at cql3/expr/expression.cc:1034:13)>, const std::variant<bool, cql3::expr::conjunction, cql3::expr::binary_operator, cql3::expr::column_value, cql3::expr::column_value_tuple, cql3::expr::token, 
cql3::expr::unresolved_identifier, cql3::expr::column_mutation_attribute, cql3::expr::function_call, cql3::expr::cast, cql3::expr::field_selection> &>' requested here
    return std::visit(overloaded_functor{
                ^
In file included from cql3/expr/expression.cc:22:
In file included from ./cql3/expr/expression.hh:27:
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/variant:1756:18: error: constexpr if condition is not a constant expression
          if constexpr (!__visit_rettypes_match)
                        ^~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/variant:1756:19: note: initializer of '__visit_rettypes_match' is not a constant expression
          if constexpr (!__visit_rettypes_match)
                         ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/variant:1752:19: note: declared here
          constexpr bool __visit_rettypes_match = __detail::__variant::
                         ^
/usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/variant:1758:22: error: static_assert expression is not an integral constant expression
              static_assert(__visit_rettypes_match,
                            ^~~~~~~~~~~~~~~~~~~~~~
```

and about one billion lines of errors follow...
```

**Comment by avikivity** [comment]
Probably needs a rebase, due to new std::visit calls.

**Comment by avikivity** [comment]
The replace_token() function gained this new cases:

```c++
            [&] (const unresolved_identifier&) -> expression {
                throw std::logic_error(format("replace_token invalid with unresolved identifier: {}", to_string(expr)));
            },
            [&] (const column_mutation_attribute&) -> expression {
                return expr;
            },
            [&] (const function_call&) -> expression {
                // A token function could be one of the arguments, but it doesn't help the caller to replace it
                // since we can't index function of the token function.
                return expr;
            },
            [&] (const cast&) -> expression {
                // A token function could be what's being casted, but it doesn't help the caller to replace it
                // since we can't index function of the token function.
                return expr;
            },
            [&] (const field_selection&) -> expression {
                return expr;
            },
```

**Comment by avikivity** [comment]
v5:
 - rebased over introduction replace_token() 
 - various patches added cases to replace_token()
 - retested

**Comment by avikivity** [comment]
@nyh ping
